core/src/main/java/hudson/PluginManager.java

```diff
@@ -1923,6 +1924,32 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
         return FormValidation.ok();
     }
 
+    @Restricted(NoExternalUse.class)
+    @RequirePOST public FormValidation doCheckUpdateSiteUrl(StaplerRequest request, @QueryParameter String value) {
+        Jenkins.get().checkPermission(Jenkins.ADMINISTER);
+        if (StringUtils.isNotBlank(value)) {
+            try {
+                value += ((value.contains("?")) ? "&" : "?") + "version=" + Jenkins.VERSION + "&uctest";
+
+                URL url = new URL(value);
+
+                // Connect to the URL
+                HttpURLConnection conn = (HttpURLConnection) ProxyConfiguration.open(url);
+                conn.setRequestMethod("HEAD");
+                conn.setConnectTimeout(5000);
+                if (100 <= conn.getResponseCode() && conn.getResponseCode() <= 399) {
+                    return FormValidation.ok();
+                } else {
+                    return FormValidation.error(Messages.PluginManager_connectionFailed());
+                }
+            } catch (IOException e) {
+                return FormValidation.error(Messages.PluginManager_connectionFailed());
```

Comment:
worth logging to help diagnosis? (untested)

```suggestion
                LOGGER.log(Level.FINE, "failed to check update site", e );
                return FormValidation.error(Messages.PluginManager_connectionFailed());
```